### PR TITLE
Xref Validation in CP_1s Commit

### PR DESCRIPTION
**Pull request by debam-das**

### CODE REVIEW

File: suites/test_xref_validation.py [hunk 1]
```diff
@@ -181,6 +181,40 @@ def xref_validation_assembly(self):
...
+
+        Xref_linkText1 = "Xref to assembly with complete path"
+        Xref_linkText2 = "xref to file on same level"
+        Xref_linkText3 = "Different module included in different assembly"
+
+        CP_title_static_part = " | Red Hat Customer Portal"
+
+        utilities.click_element(self.driver, By.LINK_TEXT, Xref_linkText1)
+        utilities.switch_to_latest_tab(self.driver)
+        check_that("Xref to assembly with complete path", utilities.get_page_title(self.driver),
```
Comment (noopurmath3) [inline suites/test_xref_validation.py:202]: matcher is missing in all the check_that statements

File: suites/test_xref_validation.py [hunk 2]
```diff
@@ -181,6 +181,40 @@ def xref_validation_assembly(self):
...
+                   "at-uploader | Assembly Publish test" + CP_title_static_part)
+        utilities.switch_to_first_tab(self.driver)
+
+        utilities.click_element(self.driver, By.LINK_TEXT, Xref_linkText2)
+        utilities.switch_to_latest_tab(self.driver)
+        check_that("xref to file on same level", utilities.get_page_title(self.driver),
+                   "Content Test Module | Logging in to the web console using Kerberos authentication (Image present)" + CP_title_static_part)
```
Comment (noopurmath3) [inline suites/test_xref_validation.py:209]: "CP_title_static_part" part is not required to be verified
You need to verify the title of the document and not the <title> tag content

File: suites/test_xref_validation.py [hunk 3]
```diff
@@ -181,6 +181,40 @@ def xref_validation_assembly(self):
...
+        utilities.switch_to_first_tab(self.driver)
+
+        utilities.click_element(self.driver, By.LINK_TEXT, Xref_linkText3)
+        utilities.switch_to_latest_tab(self.driver)
```
Comment (noopurmath3) [inline suites/test_xref_validation.py:213]: Add all the code that leads to opening a new tab in a try blocks
and at the end add a finally block which navigates to the main tab after the execution is completed
Refer existing tests for reference

**Review by noopurmath3 — CHANGES_REQUESTED** [review]
Please execute this individual test and then the entire suite once before creating/updating the PR

File: helpers/constants.py [hunk 4]
```diff
@@ -138,4 +135,12 @@
 # xref validation constants
 partial_title = "xref validation"
 xref_validation_msg = "Invalid cross references"
-
+xref_validation_module_name = "at-uploader | This doc contains a list of refs"
+# xref validation constants in CP
+Xref_linkText1 = "at-uploader | Assembly Publish test"
```
Comment (noopurmath3) [inline helpers/constants.py:140]: No need to have individual constants and the again a dict containing the same values
You can access individual values from the dict itself

File: helpers/locators.py [hunk 5]
```diff
@@ -52,9 +52,10 @@
 XREF_VALIDATION_COUNT_CSS = "li#xrefs>div>button span.pf-c-badge.pf-m-read"
 PATH_TO_ADOC_CSS = "div.pf-l-level>div>div.pf-c-content pre"
 GENERATE_HTML_CSS = "button.pf-c-button.pf-m-secondary"
-
+CUSTOMER_PORTAL_XPATH = "//a[text()='View on Customer Portal  ']"
 # Customer portal
 MODULE_TITLE_ON_PORTAL_ID= "rhdocs-header"
+MODULE_TITLE_XPATH= "//header[@class='rhdocs__header']//h1"
```
Comment (noopurmath3) [inline helpers/locators.py:58]: CSS locator should be preferred over xpath

File: helpers/locators.py [hunk 6]
```diff
@@ -74,6 +75,7 @@
 LEGAL_NOTICE_ON_PORTAL_CSS = "article#rhdocs-content div.rh-docs-legal-notice a.rh-docs-legal-notice__link"
 CONTENT_RELATED_GUIDES_RESOURCES = "details.related-topic-content__wrapper--for-guide aside div.sectionbody"
 CONTENT_RELATED_TO_GUIDES = "details.related-topic-content__wrapper--for-guide"
+CONSENT_BUTTON_XPATH = "//button[@class='truste-button1']"
```
Comment (noopurmath3) [inline helpers/locators.py:78]: CSS locator should be preferred over xpath

**Review by noopurmath3 — CHANGES_REQUESTED** [review]
@debam-das 
- All the comments added in the previous round of review haven't been addressed yet
- Once a fix/change has been made to address a comment, mark that comment as Resolved from the gitlab UI after pushing the changes
- Also, I can see that there are a lot of changes made, which mostly look like have been made because of the IDE config. Please revert those changes and update the IDE config as well, as it is very confusing to review the actual part that has been updated
- Along with that please remove all the empty lines that have been added after evey statement, eg after line 200 in test_xref_validation.py there is an empty line after almost every statement, which is not required

File: suites/test_xref_validation.py (hunk 2 above)
Comment (debam-das) [reply suites/test_xref_validation.py]: Done. Xref Validation in CP_3rd Commit

File: suites/test_xref_validation.py (hunk 1 above)
Comment (debam-das) [reply suites/test_xref_validation.py]: Done.Xref Validation in CP_3rd Commit

File: suites/test_xref_validation.py [hunk 7]
```diff
@@ -181,6 +181,40 @@ def xref_validation_assembly(self):
...
+    def xref_validations_module_in_CP(self):
+        utilities.click_element(self.driver, By.LINK_TEXT, locators.MENU_SEARCH_PAGE_LINK_TEXT)
+        utilities.page_reload(self.driver)
+        search_beta_page.select_repo(self.driver, fixture.repo_name)
+        search_beta_page.search_module_and_click(self.driver, constants.xref_validation_module_name)
+        utilities.wait(4)
+        display_module_page.add_metadata_and_publish(self.driver)
+        display_module_page.navigate_into_CP(self.driver)
+
+        Xref_linkText1 = "Xref to assembly with complete path"
```
[the remark this answers is not in the document]
Comment (debam-das) [reply suites/test_xref_validation.py]: Done.Xref Validation in CP_3rd Commit

File: suites/test_xref_validation.py (hunk 3 above)
Comment (debam-das) [reply suites/test_xref_validation.py]: Done. Xref Validation in CP_3rd Commit

File: helpers/locators.py [hunk 8]
```diff
@@ -52,7 +52,7 @@
 XREF_VALIDATION_COUNT_CSS = "li#xrefs>div>button span.pf-c-badge.pf-m-read"
 PATH_TO_ADOC_CSS = "div.pf-l-level>div>div.pf-c-content pre"
 GENERATE_HTML_CSS = "button.pf-c-button.pf-m-secondary"
-
+CUSTOMER_PORTAL_XPATH = "//a[text()='View on Customer Portal  ']"
```
[the remark this answers is not in the document]
Comment (debam-das) [reply helpers/locators.py:55]: Done.

File: helpers/utilities.py [hunk 9]
```diff
@@ -71,13 +71,15 @@ def find_elements_by_css_selector(driver, locator):
         lcc.log_error(str(e))
     return elements_list_by_css_selector
 
+
```
[the remark this answers is not in the document]
Comment (debam-das) [reply helpers/utilities.py]: Done.

File: helpers/utilities.py [hunk 10]
```diff
@@ -120,22 +122,37 @@ def generate_random_string(string_length):
     random_string = ''.join(random.choice(letters) for i in range(string_length))
     return random_string
 
+
```
[the remark this answers is not in the document]
Comment (debam-das) [reply helpers/utilities.py]: Done.

File: helpers/utilities.py [hunk 11]
```diff
@@ -120,22 +122,37 @@ def generate_random_string(string_length):
...
-        shadow_host = find_element(driver, By.CSS_SELECTOR,shadow_root_parent )
+        shadow_host = find_element(driver, By.CSS_SELECTOR, shadow_root_parent)
     except (TimeoutException, NoSuchElementException) as e:
         lcc.log_error(e)
     else:
         root_ele = driver.execute_script("return arguments[0].shadowRoot", shadow_host)
         return root_ele
 
-def find_shadow_dom_element (driver,locator, shadow_root_parent):
+
```
[the remark this answers is not in the document]
Comment (debam-das) [reply helpers/utilities.py]: Done.

File: helpers/utilities.py [hunk 12]
```diff
@@ -120,22 +122,37 @@ def generate_random_string(string_length):
...
+def find_shadow_dom_element(driver, locator, shadow_root_parent):
     shadow_root = get_shadow_root(driver, shadow_root_parent)
     element = shadow_root.find_element_by_css_selector(locator)
     return element
 
+
```
[the remark this answers is not in the document]
Comment (debam-das) [reply helpers/utilities.py]: Done.

File: helpers/utilities.py [hunk 13]
```diff
@@ -71,13 +71,15 @@ def find_elements_by_css_selector(driver, locator):
...
 def find_elements_by_XPATH(driver, locator):
     try:
         elements_list_by_XPATH = driver.find_elements(By.XPATH, locator)
     except NoSuchElementException as e:
         lcc.log_error(str(e))
     return elements_list_by_XPATH
 
+
```
[the remark this answers is not in the document]
Comment (debam-das) [reply helpers/utilities.py]: Done.

File: suites/test_xref_validation.py [hunk 14]
```diff
@@ -181,6 +181,40 @@ def xref_validation_assembly(self):
                 check_that("File content", content, contains_string(target_path))
             i = i + 1
 
+    @lcc.test('Verify xref validation for module in Customer Portal')
+    def xref_validations_module_in_CP(self):
+        utilities.click_element(self.driver, By.LINK_TEXT, locators.MENU_SEARCH_PAGE_LINK_TEXT)
+        utilities.page_reload(self.driver)
```
[the remark this answers is not in the document]
Comment (debam-das) [reply suites/test_xref_validation.py]: I also observed same thing, Removed it.

File: helpers/constants.py [hunk 15]
```diff
@@ -140,10 +140,13 @@
 xref_validation_msg = "Invalid cross references"
 xref_validation_module_name = "at-uploader | This doc contains a list of refs"
 # xref validation constants in CP
+jseq_linkText1 = "return document.querySelector('#rhdocs').shadowRoot.querySelector('p:nth-child(3) > a').click()"
```
Comment (noopurmath3) [inline helpers/constants.py:143]: @debam-das 
Please move these 3 lines to locators file

**Review by noopurmath3 — APPROVED** [review]
@debam-das Please address the 1 comment 
Everything else LGTM 👍

**Pull request merged**